src/logging.h

```diff
@@ -37,41 +37,109 @@ struct LogCategory {
 };
 
 namespace BCLog {
-    enum LogFlags : uint32_t {
-        NONE        = 0,
-        NET         = (1 <<  0),
-        TOR         = (1 <<  1),
-        MEMPOOL     = (1 <<  2),
-        HTTP        = (1 <<  3),
-        BENCH       = (1 <<  4),
-        ZMQ         = (1 <<  5),
-        WALLETDB    = (1 <<  6),
-        RPC         = (1 <<  7),
-        ESTIMATEFEE = (1 <<  8),
-        ADDRMAN     = (1 <<  9),
-        SELECTCOINS = (1 << 10),
-        REINDEX     = (1 << 11),
-        CMPCTBLOCK  = (1 << 12),
-        RAND        = (1 << 13),
-        PRUNE       = (1 << 14),
-        PROXY       = (1 << 15),
-        MEMPOOLREJ  = (1 << 16),
-        LIBEVENT    = (1 << 17),
-        COINDB      = (1 << 18),
-        QT          = (1 << 19),
-        LEVELDB     = (1 << 20),
-        VALIDATION  = (1 << 21),
-        I2P         = (1 << 22),
-        IPC         = (1 << 23),
+    enum LogFlags {
+        NET,
+        TOR,
+        MEMPOOL,
+        HTTP,
+        BENCH,
+        ZMQ,
+        WALLETDB,
+        RPC,
+        ESTIMATEFEE,
+        ADDRMAN,
+        SELECTCOINS,
+        REINDEX,
+        CMPCTBLOCK,
+        RAND,
+        PRUNE,
+        PROXY,
+        MEMPOOLREJ,
+        LIBEVENT,
+        COINDB,
+        QT,
+        LEVELDB,
+        VALIDATION,
+        I2P,
+        IPC,
 #ifdef DEBUG_LOCKCONTENTION
-        LOCK        = (1 << 24),
+        LOCK,
 #endif
-        BLOCKSTORAGE = (1 << 25),
-        TXRECONCILIATION = (1 << 26),
-        SCAN        = (1 << 27),
-        TXPACKAGES  = (1 << 28),
-        ALL         = ~(uint32_t)0,
+        BLOCKSTORAGE,
+        TXRECONCILIATION,
+        SCAN,
+        TXPACKAGES,
+        // Add new entries before this line, and also add a
+        // corresponding entry to LOG_CATEGORIES_BY_STR.
+
+        // The following have no representation in m_categories:
+        ALL, // this is also the size of the bitset
+        NONE,
     };
+
+    template<typename FlagType, size_t BITS, typename T>
+    class AtomicBitSet {
+    private:
+        // bits in a single T (an integer type)
+        static constexpr size_t BITS_PER_T = 8 * sizeof(T);
+
+        // number of Ts needed for BITS bits (round up)
+        static constexpr size_t NT = (BITS + BITS_PER_T - 1) / BITS_PER_T;
+
+        std::atomic<T> bits[NT]{0};
+
+    public:
+        AtomicBitSet() = default;
+        AtomicBitSet(FlagType f) { set(f); }
+
+        AtomicBitSet(const AtomicBitSet&) = delete;
+        AtomicBitSet(AtomicBitSet&&) = delete;
+        AtomicBitSet& operator=(const AtomicBitSet&) = delete;
+        AtomicBitSet& operator=(AtomicBitSet&&) = delete;
+        ~AtomicBitSet() = default;
+
+        constexpr size_t size() const { return BITS; }
+
+        bool any() const
+        {
+            for (const auto& i : bits) {
+                if (i > 0) return true;
+            }
+            return false;
+        }
```

Comment:
This can observe a state that never was. For example, lets say `bits[]` has 2 elements, initially `[0000, 0010]`:

0. `[0000, 0010]`, `any() == true`
1. `any()` is called and it reads the first element `0000`
2. another thread sets the 1st bit, new state: `[1000, 0010]`, `any() == true`
3. another thread resets the 7th bit, new state: `[1000, 0000]`, `any() == true`
4. the execution of `any()` from 1. continues and reads the second element: `0000` and that `any()` call returns `false`, even though the bitset was never empty.

I would prefer the dump solution of changing to 64 bits from https://github.com/bitcoin/bitcoin/pull/26619. ACK 90c3d904bcbf294a6fcef8bc9ecb9445fd41c4bd from that PR (I can't post it there because it is locked). 64 would be sufficient for a loooong time and personally, I don't see "I think there's no urgent need for >32 bits, so might as well go straight to bitset."[*](https://github.com/bitcoin/bitcoin/pull/26619#issuecomment-1344545801) as enough justification to jump into over-engineering this.

Reply:
+1 for uint64_t instead of AtomicBitSet

Reply:
As I look at this again, I agree this is over-engineered (what do you think, @ryanofsky?). #26619 was my first attempt but someone suggested using `std::bitset` and I took that suggestion. But that approach (this PR)  turned out to be more complicated than expected.

The one thing I like about the current approach is not having to specify all the bit shift values explicitly, such as `NET = (1 <<  0)`, instead it's just a clean list. But that isn't much of an advantage.

Why don't I rebase #26619 and get it re-opened (I think a maintainer has to do that, @ryanofsky?), then both will be open and up to date, and reviewers can decide which to ACK (or maybe ACK both and a maintainer decides). I don't mind either one, but slightly prefer #26619 because it is simpler, and 64 categories should last a very long time. In the worst case, if we need even more, this PR can be resurrected.

Reply:
Hard to say if it's a bug, it's definitely not undefined behavior. This `any()` function is currently [used](https://github.com/bitcoin/bitcoin/pull/26697/files#diff-2ab59d52a34bfed0bded8d7c7ad995367b95dbb983b29efca5d2f56a59fa3ccbL339-L340) only by `qt` where it says, if there are any debug logs enabled, then display a bunch of extra debug stuff (which is a bit questionable, because the logging categories that are enabled may have nothing whatsoever to do with `qt`). So if the sequence vasild outlined above happens, and the logging is being changed when `qt` is deciding whether to show this debug stuff, then it could make the wrong decision. But it's just debug stuff.

Of course, there could be other calls to `any()` in the future.

The nature of a function like `any()` is that as soon as you call it, it could change anyway. But what vasild found is slightly worse, because there was never a moment where no logging was enabled, yet `any()` returns false.

I tried to fix the problem but it turns out to be quite messy. I suppose this is another reason to go with the other approach (`uint64_t`), although not a strong reason, IMO.

Thanks for reopening #26619, Marco. I'll rebase it now.

Reply:
This is a good catch and I think it's worth commenting on, but I don't think it is a bug. Would suggest adding a comment like: "Checks if any bits are set. Note: Bits are checked independently, so result reflects the state of all bits, but not at a single instant in time."